Workaround for IPv6 EPMD

Gsantomaggio opened the pull request:
Add workaround for EPMD IPV6.

## Proposed Changes

Possible workaround / fix for [ipv6 epmd issue](https://groups.google.com/forum/#!searchin/rabbitmq-users/ipv6%7Csort:date/rabbitmq-users/mbsT8PSJTHQ/tjAgsRULBgAJ)
(Until Erlang won't fix the [issue](https://github.com/erlang/otp/pull/602)) 

Had to add:
```
 ${RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS} \
 ``` 
 In `rabbit_prelaunch` since it wans't not possible to [pass parameters](https://github.com/rabbitmq/rabbitmq-server/blob/v3.7.x/scripts/rabbitmq-server#L191)
 
 in this case:
 
 ```
 RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS = " -proto_dist inet6_tcp "
 RABBITMQ_CTL_ERL_ARGS = " -proto_dist inet6_tcp "
 ```
 
The enviroment where i tested it is the same I described [here](https://groups.google.com/forum/#!searchin/rabbitmq-users/ipv6%7Csort:date/rabbitmq-users/mbsT8PSJTHQ/tjAgsRULBgAJ)
 And with this patch I can start the rabbitmq-server.
 
 
 
if you want to test it without rabbitmq it is enough to:
```
erl -proto_dist inet6_tcp  -sname hello
net_adm:names(localhost).
{error,address}
```

Try again using:

```
erl -proto_dist inet6_tcp  -sname hello
net_adm:names({0,0,0,0,0,0,0,1}).
{ok,[{"hello",36371}]}
```

_In short this fix tries to use the local IPV6 address when for some reason the `net_adm:names(..)` fails with the default value_
 
## Types of Changes

- [X] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist
- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

**NOTE** this is just an idea, the code should be organised differnetly as for example the variable:
 ```
 -define(LOCAL_IPV6_ADDRESS, {0,0,0,0,0,0,0,1}).
``` 
should be inside some common file.

Let me know what you think.

Thank you

michaelklishin commented:
Thank you for looking into it!

I wonder if it's just my unfounded fears or using `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` with prelaunch could be dangerous? We can introduce `RABBITMQ_PRELAUNCH_ADDITIONAL_ERL_ARGS` to limit the scope/eliminate unintentional reuse.

michaelklishin commented:
@dumbbell do you have an opinion about a new variable just for `prelaunch`?

Gsantomaggio commented:
I was wondering the same, I'd tend to add a new var:
`RABBITMQ_PRELAUNCH_ADDITIONAL_ERL_ARGS`

lukebakken commented:
> I wonder if it's just my unfounded fears or using RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS with prelaunch could be dangerous?

I think we _should_ use `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` with prelaunch. If a bad flag is passed we should catch it early.

dumbbell commented on src/rabbit_epmd_monitor.erl:
```diff
@@ -27,6 +27,8 @@
 
 -define(SERVER, ?MODULE).
 -define(CHECK_FREQUENCY, 60000).
+-define(LOCAL_IPV6_ADDRESS, {0,0,0,0,0,0,0,1}).
```
I believe `LOOPBACK` would be more accurate than `LOCAL` in the name, because any IP address assigned to this host is local.

dumbbell commented on src/rabbit_epmd_monitor.erl:
```diff
@@ -49,10 +51,17 @@
+
+epmd_port_please(Mod, Name, Host)->
+  case Mod:port_please(Name, Host) of
+    noport -> erl_epmd:port_please(Name, ?LOCAL_IPV6_ADDRESS);
```
Why use `Mod` for the first call, then hard-code `erl_epmd`? I suppose this is a copy-paste error.

dumbbell replied:
Also, would it be possible to move this function to a single module, export it and use it everywhere, including `rabbit_networking` and so on? Plus a comment explaining why we need this.

Gsantomaggio replied:
> Why use `Mod` for the first call, then hard-code `erl_epmd`? I suppose this is a copy-paste error.

Ops, yes, changed :)! 

>Also, would it be possible to move this function to a single module

Ok done

dumbbell commented on src/rabbit_prelaunch.erl:
```diff
@@ -151,11 +153,24 @@ dist_port_use_check_ipv6(NodeHost, Port) ->
 
 dist_port_use_check_fail(Port, Host) ->
-    {ok, Names} = rabbit_nodes:names(Host),
+    io:format("dist_port_use_check_fail port:~p host:~p~n",[Port, Host]),
```
Is it a debug message left over?

dumbbell commented on src/rabbit_prelaunch.erl:
```diff
@@ -151,11 +153,24 @@ dist_port_use_check_ipv6(NodeHost, Port) ->
+       {error, EpmdReason} ->
+           io:format("ERROR: epmd error for host ~s: ~s, trying with ipv6~n",
+               [NodeHost, rabbit_misc:format_inet_error(EpmdReason)]),
+           rabbit_nodes:names(?LOCAL_IPV6_ADDRESS);
```
Could you please add a comment explaning the problem solved/worked around here?

Gsantomaggio replied:
@dumbbell did it.
I left `empd_names(NodeHost)` inside  `rabbit_prelaunch.erl`. Please let me know if you want to move also this function to  `rabbit_networking`

dumbbell commented:
The prelaunch node and the final node should probably be started with the same flags to reproduce the same conditions. So I'm leaning toward reusing `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` as @lukebakken suggests, but I understand @michaelklishin's fear.

Do we have some documentation about the actual mission of the prelaunch node?

michaelklishin commented:
prelaunch is an implementation detail that was originally added to make sure that `epmd` is started IIRC. Now it performs a bit of sanity checking of various startup arguments. So @lukebakken's idea makes sense.

lukebakken commented:
@dumbbell [code](https://github.com/rabbitmq/rabbitmq-server/blob/master/src/rabbit_prelaunch.erl#L35-L51) - Checks to see if a duplicate node is running, and also is used to parse config and do various dist port checks.

dumbbell commented:
Ok, thank you. As I started to study rabbitmq-server(8) with my idea to move everything, to Erlang code (in the context of the final Erlang node), that prelaunch node is part of that topic. Perhaps we can start the final node without distribution turned on and do what that prelaunch does during RabbitMQ boot (as well as activating distribution).

Anyway, it's out of scope for this particular issue.

Gsantomaggio commented:
I created and[ Vagrant image](https://github.com/Gsantomaggio/rabbitmq-utils/blob/master/rabbitmq-suse/leap_15_ipv6/Vagrantfile) to help to test the issue.

1.  setup 
```bash
vagrant up && vagrant ssh
```

2. Try to start rabbitmq
```bash
:~> sudo ss -tulpen | grep epmd
tcp    LISTEN     0      128       [::1]:4369               [::]:*                   users:(("epmd",pid=1072,fd=3),("systemd",pid=1,fd=46)) ino:15225 sk:4 v6only:1 <->

:~> sudo systemctl start rabbitmq-server
Job for rabbitmq-server.service failed because the control process exited with error code.
See "systemctl  status rabbitmq-server.service" and "journalctl  -xe" for details.
```

3.  Add ipv6 distro support, and try again 
```bash 
:~> sudo bash -c  'echo RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS=\"-proto_dist inet6_tcp \" > /etc/rabbitmq/rabbitmq-env.conf'
:~> sudo bash -c  'echo RABBITMQ_CTL_ERL_ARGS="-proto_dist inet6_tcp \" >> /etc/rabbitmq/rabbitmq-env.conf'

:~> sudo systemctl start rabbitmq-server
:~>sudo journalctl  -xe
...
-- Unit rabbitmq-server.service has begun starting up.
apr 18 11:30:36 localhost rabbitmq-server[9892]: Protocol 'inet_tcp': register/listen error: econnrefused
apr 18 11:30:36 localhost systemd[1]: rabbitmq-server.service: Main process exited, code=exited, status=1/FAILURE
apr 18 11:30:36 localhost systemd[1]: Failed to start RabbitMQ broker.
``` 

4. Apply the patch:
```bash
~> sudo cp /vagrant/*.beam /usr/lib/rabbitmq/lib/rabbitmq_server-3.8.0-beta.3/ebin/
~> sudo cp /vagrant/rabbit /usr/lib/rabbitmq/lib/rabbitmq_server-3.8.0-beta.3/sbin/
``` 

5. Start rabbitmq again:

```bash
vagrant@localhost:~> sudo systemctl start rabbitmq-server
vagrant@localhost:~> sudo rabbitmqctl status
Status of node rabbit@localhost ...
[{pid,4049},
 {running_applications,
     [{rabbit,"RabbitMQ","3.8.0-beta.3"},
      {ra,"Experimental Raft library","0.9.3"},
      {aten,"Erlang node failure detector","0.5.2"},
      {sysmon_handler,"Rate-limiting system_monitor event handler","1.1.0"},
      {os_mon,"CPO  CXC 138 46","2.4.7"},
```

spacefito commented:
For what is worth.  I have found that when trying to deploy rabbitmq cluster on ipv6 only hosts. 
It is best to pre-install elrang (epmd and epmd.socket) *and* configuring epmd.socket to listen on ipv6 only *before* installing and starting rabbitmq.

michaelklishin commented:
Please limit this discussion to this specific PR. Once RabbitMQ is started, so will be `epmd` and any configuration applied to it after the fact won't have any effect.

Gsantomaggio commented:
Hi all,
Sorry for the delay, but it seems the erlang PR ( otp/pull/2222 ) requires time.

I replaced the  `-define(LOOPBACK_IPV6_ADDRESS, {0,0,0,0,0,0,0,1}).`  with `inet:getaddr(Host, inet6)` that is more correct.

Also added more comments.

Thanks!

michaelklishin commented:
@Gsantomaggio no worries. Should we put this one pause until erlang/otp#2222 is in?

Gsantomaggio commented:
I would prefer to merge it so we won't depend on the Erlang version. 
When/if the otp/2222 will be merged we will decide.

Gsantomaggio commented:
I tested the way suggested by Lukas here https://github.com/erlang/otp/pull/2222#issuecomment-496923946

```
 cat /etc/rabbitmq/rabbitmq-env.conf
RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="-kernel inetrc '/etc/rabbitmq/erl_inetrc'  -proto_dist inet6_tcp "
RABBITMQ_CTL_ERL_ARGS="-proto_dist inet6_tcp "
```
and RabbitMQ starts, only with **[rabbitmq-server](https://github.com/rabbitmq/rabbitmq-server/pull/1982/files#diff-506d97606d1e74d158f8624a8cc9e091R197)** fix. ( Without the other erlang files)

So, less impact and less problems.

have to think about the Lukas question:
>Can you think of any reason why you would want to resolve ipv4 addresses per default and use ipv6 distro?

With my currect fix, we try both, with `{inet6,true}.` we have everything in ipv6. 
Thoughts?

michaelklishin commented:
@Gsantomaggio if it works in your environment I see no reason not to recommend it. No code changes is a pretty great outcome for this problem if you ask me :)

michaelklishin commented:
It's worth mentioning that @dumbbell is working on moving most of `rabbitmq-server` logic to Erlang. However, we intend to support the same set of env variables, and one added to make the `erl_inetrc` option work is a reasonable short term compromise.

Gsantomaggio commented:
I tried it on my test enviroment and it works, we will try it in our systems asap and will let you know. 
Thank you for the patience in this issue, but I didn't know all these details.

>@dumbbell is working on moving most of rabbitmq-server logic to Erlang. 

That's a very good news, it will be easier to handle.

Gsantomaggio commented:
@michaelklishin @lukebakken  that's it :)! 
wondering if we have to add a section on the documentation about `IPV6` and `erl_inetrc` configuration

michaelklishin commented:
@Gsantomaggio yes, a section on IPv6 only environments in the networking guide makes sense to me.

lukebakken commented:
Backported to `v3.7.x`
